plugins/producer_plugin/include/eosio/producer_plugin/snapshot_scheduler.hpp

```diff
@@ -0,0 +1,207 @@
+                                                 BOOST_MULTI_INDEX_MEMBER(producer_plugin::snapshot_request_information, uint32_t, end_block_num)>>>>;
+   snapshot_requests _snapshot_requests;
+   snapshot_db_json _snapshot_db;
+   uint32_t _snapshot_id   = 0;
+   uint32_t _inflight_sid  = 0;
+   std::function<void(producer_plugin::next_function<producer_plugin::snapshot_information>)> _create_snapshot;
+
+   void x_serialize() {
+      auto& vec = _snapshot_requests.get<as_vector>();
+      std::vector<producer_plugin::snapshot_schedule_information> sr(vec.begin(), vec.end());
+      _snapshot_db << sr;   
+   }
+
+public:
+   snapshot_scheduler() = default;
+
+   // snapshot_scheduler_listener
+   void on_start_block(uint32_t height) {
+      bool serialize_needed  = false;
+      bool snapshot_executed = false; 
+
+      auto execute_snapshot_with_log = [this, &height, &snapshot_executed](const auto & req) {
+         // one snapshot per height
+         if (!snapshot_executed) {
+            dlog("snapshot scheduler creating a snapshot from the request [start_block_num:${start_block_num}, end_block_num=${end_block_num}, block_spacing=${block_spacing}], height=${height}",
+               ("start_block_num", req.start_block_num)
+               ("end_block_num",   req.end_block_num)
+               ("block_spacing",   req.block_spacing)
+               ("height",          height));
+               
+            execute_snapshot(req.snapshot_request_id);
+            snapshot_executed = true;
+         }
+      };    
+
+      for(const auto& req: _snapshot_requests.get<0>()) {
+         // -1 since its called from start block
+         bool recurring_snapshot =  req.block_spacing &&  (height >= req.start_block_num + 1) && (!((height - req.start_block_num - 1) % req.block_spacing));
+         bool onetime_snapshot   = (!req.block_spacing) && (height == req.start_block_num + 1);
+        
+         // assume "asap" for snapshot with missed/zero start, it can have spacing
+         if(!req.start_block_num) {
+            // update start_block_num with current height only if this is recurring
+            // if non recurring, will be executed and unscheduled
+            if (req.block_spacing && height) {
+               auto& snapshot_by_id = _snapshot_requests.get<by_snapshot_id>();
+               auto it = snapshot_by_id.find(req.snapshot_request_id);
+               _snapshot_requests.modify(it, [&height](auto& p) { p.start_block_num = height - 1; });
+               serialize_needed = true;              
+            }
+            execute_snapshot_with_log(req);
+         }        
+         else if(recurring_snapshot || onetime_snapshot) {
+            execute_snapshot_with_log(req);
+         }       
+
+         // cleanup - remove expired (or invalid) request
+         if(!req.start_block_num ||         
```

Comment:
This check `(!req.start_block_num || ` will remove all recurring snapshot requests which have `start_block_num == 0`, even if they are not expired. My version with two loops fixed that.

Reply:
I am pretty sure at this line req.start_block_num==0 should not ever happen, being corrected few lines above, but I agree, your suggestion is safest one! Thank you!

Reply:
> I am pretty sure at this line req.start_block_num==0 should not ever happen, being corrected few lines above, but I agree, your suggestion is safest one! Thank you!

I wasn't sure, because I see we update the request in boost multi-index, but `req` is a `const &`  to a view of that multi-index and I was not sure it was necessarily updated.